### ISSUE TITLE
Syntax coloring fails

### ISSUE THREAD

**Issue by odysseus55** [atuttle/atom-language-cfml#120]
Just upgraded to language-cfml 0.23.0 and the syntax coloring stops at various place in different files. I have attached a screenshot.

**Comment by David-Polehonski** [comment]
Could you provide a test case/file where the highlighting fails predictably? Once I have a test case I'll get working on a fix.

**Comment by odysseus55** [comment]
HI!

Thanks for getting back to me on this. I have had some time to fiddle with
the issue and hopefully can point you in the right direction.
Attached is a .cfm file zipped. It has 238 lines, mostly cfml but some
javascript. This is where the error is exhibited.

To see the error, uncomment lines 158 and 162.

In the javascript I have some CF variables that need to be extracted and
inserted into the JS . The problem is with the <cfoutput> tag. (see code
attached)

Line 158 I have commented out the cfoutput tag (the highlighting is working
properly). But I do need to extract the CF vars, I would prefer to bracket
just the relevant lines as shown (lns 158 & 162) rather than the whole
<script> section. If I move the <cfoutput> to outside the <script> then the
highlighting works well, but then I need to escape every other # in the
section. This is, of course, a real pain.

Thanks for all you do on this module. I have used it for years and it makes
life easier.

Let me know if you want anything else.


Dale

**Comment by David-Polehonski** [comment]
That description is very helpful, and gives me a good place to get started - unfortunately your .zip didn't come through. If you could send it through as gist, or PM  if you prefer the privacy that'd be great.

**Comment by odysseus55** [comment]
<cfsilent>
</cfsilent>
<html class="no-js" lang="en">
<head>
	<title>LimoZone</title>
	<meta name="application-name" content="Zone" />
	<meta name="msapplication-config" content="ieconfig.xml" />
	<link rel="shortcut icon" href="/images/favicon.ico" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
    <meta charset="utf-8">
    <meta name="ROBOTS" content="NOINDEX, NOFOLLOW">
	<meta http-equiv="cache-control" content="no-cache" />
	<meta http-equiv="expires" content="Tue, 01 Jan 1980 1:00:00 GMT" />
	<meta http-equiv="pragma" content="no-cache" />
	<link rel="stylesheet" href="/css/stylesheet.css" type="text/css" />

	<link rel="stylesheet" href="/css/limozone.css" type="text/css" />
	<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jquery/2.1.3/jquery.min.js"></script>
	<link href='https://fonts.googleapis.com/css?family=Oxygen:300,400,700|Oswald:300,400,700' rel='stylesheet' type='text/css'>
<style>
table {
    white-space: normal;
    line-height: normal;
    font-weight: normal;
    font-size: 24px;
    font-style: normal;
	width:500px;
	-webkit-border-radius: 4;
	-moz-border-radius: 4;
	border-radius: 4px;
}
input[type=number]{
    width: 70px;
}
.center {
	margin: auto;
	width: 500px;
	padding: 10px;
}
.bigbtn {
  background: #37e65d;
  background-image: -webkit-linear-gradient(top, #7F9591, #C7D6C8 );
  background-image: -moz-linear-gradient(top, #7F9591, #C7D6C8);
  background-image: -ms-linear-gradient(top, #7F9591, #C7D6C8);
  background-image: -o-linear-gradient(top, #7F9591, #C7D6C8);
  background-image: linear-gradient(to bottom, #7F9591, #C7D6C8);
  -webkit-border-radius: 4;
  -moz-border-radius: 4;
  border-radius: 4px;
  text-shadow: 1px 1px 3px #666666;
  font-family: Arial;
  text-transform: uppercase;
  color: #000;
  width:165px;
  font-size: 24px;
  padding: 5px 20px 5px 20px;
  text-decoration: none;
}

.bigbtn:hover {
  background: #08a32a;
  background-image: -webkit-linear-gradient(top, #C7D6C8, #7F9591);
  background-image: -moz-linear-gradient(top, #C7D6C8, #7F9591);
  background-image: -ms-linear-gradient(top, #C7D6C8, #7F9591);
  background-image: -o-linear-gradient(top, #C7D6C8, #7F9591);
  background-image: linear-gradient(to bottom, #C7D6C8, #7F9591);
  width:165px;
  text-decoration: none;
}


.bigbtn:selected {
  background: #98a32a;
  background-image: -webkit-linear-gradient(top, #9cb0fd, #3498db);
  background-image: -moz-linear-gradient(top, #9cb0fd, #3498db);
  background-image: -ms-linear-gradient(top, #9cb0fd, #3498db);
  background-image: -o-linear-gradient(top, #9cb0fd, #3498db);
  background-image: linear-gradient(to bottom, #9cb0fd, #3498db);
  text-decoration: underline;
}
#foo {
    position: fixed;
    bottom: 10;
    right: 10;
}
#closeArea, #altbutton {
	display:none;
}
</style>
</head>
<body>
	<div id="wrapper">
		<div id="getArrived" class="center biglines">
			<h1 id="arrived_id">Guest Counts</h1>
			<div id="tablearea">
				<table>
					<tr><td align="right">Locals?&nbsp;</td><td><input id = "paxArriveLocalA" size="5" value="" style="line-height:2.5em; font-size:24px;" type="number" step="1">
						&nbsp;&nbsp;&nbsp;
						<select name="paxArriveLocalB" id="paxArriveLocalB" class="bigbtn">
							<option value="">&nbsp;select</option>
							<option value="1">&nbsp;1</option>
							<option value="2">&nbsp;2</option>
							<option value="3">&nbsp;3</option>
							<option value="4">&nbsp;4</option>
							<option value="5">&nbsp;5</option>
							<option value="6">&nbsp;6</option>
							<option value="7">&nbsp;7</option>
							<option value="8">&nbsp;8</option>
							<option value="9">&nbsp;9</option>
							<option value="10">&nbsp;10</option>
						</select>
					</td></tr>
					<tr><td align="right">Non Locals?&nbsp;</td><td><input id = "paxArriveNonLocalA" size="5" value="" style="line-height:2.5em; font-size:24px;" type="number" step="1">
						&nbsp;&nbsp;&nbsp;
						<select name="paxArriveNonLocalB" id="paxArriveNonLocalB" class="bigbtn">
							<option value="">&nbsp;select</option>
							<option value="1">&nbsp;1</option>
							<option value="2">&nbsp;2</option>
							<option value="3">&nbsp;3</option>
							<option value="4">&nbsp;4</option>
							<option value="5">&nbsp;5</option>
							<option value="6">&nbsp;6</option>
							<option value="7">&nbsp;7</option>
							<option value="8">&nbsp;8</option>
							<option value="9">&nbsp;9</option>
							<option value="10">&nbsp;10</option>
						</select>
					</td></tr>
					<tr><td align="right">International?&nbsp;</td><td><input id = "paxArriveInternationalA" size="5" value="" style="line-height:2.5em; font-size:24px;" type="number" step="1">
						&nbsp;&nbsp;&nbsp;
						<select name="paxArriveInternationalB" id="paxArriveInternationalB" class="bigbtn">
							<option value="">&nbsp;select</option>
							<option value="1">&nbsp;1</option>
							<option value="2">&nbsp;2</option>
							<option value="3">&nbsp;3</option>
							<option value="4">&nbsp;4</option>
							<option value="5">&nbsp;5</option>
							<option value="6">&nbsp;6</option>
							<option value="7">&nbsp;7</option>
							<option value="8">&nbsp;8</option>
							<option value="9">&nbsp;9</option>
							<option value="10">&nbsp;10</option>
						</select>
					</td></tr>
					<tr><td id="thecell" align="center" colspan="3"><button id="mainbutton" class="bigbtn close" onClick="doCabDrop()"> Done </button><button  id="altbutton" class="bigbtn close " style="color: red;"> Saved </button></td></tr>
				</table>
			</div>

		</div>
		<div><button  id="view-fullscreen" >Full Screen</button><button  id="backmenu" >Main Menu</button></div>
	
	</div>
<cfinclude template="footer.cfm">

<script type="text/javascript">
	function doCabDrop(){
		var params = new Object();
			<cfoutput>
			params.pudate = '#dateformat(now(),"yyyy-mm-dd")#';
			params.putime = '#timeformat(now(),"HH:nn:ss")#';
			params.userid = '#session.googleid#';
			</cfoutput>

			var p1 = ($( '#paxArriveLocalA' ).val() != '') ? $( '#paxArriveLocalA' ).val(): $( '#paxArriveLocalB' ).val();
			var p2 = ($( '#paxArriveNonLocalA' ).val() != '') ? $( '#paxArriveNonLocalA' ).val(): $( '#paxArriveNonLocalB' ).val();
			var p3 = ($( '#paxArriveInternationalA' ).val() != '') ? $( '#paxArriveInternationalA' ).val(): $( '#paxArriveInternationalB' ).val();
			if (p1==''){p1=0};
			if (p2==''){p2=0};
			if (p3==''){p3=0};
			params.pax = parseInt(p1)+parseInt(p2)+parseInt(p3);
			params.locals = p1;
			params.nonlocal=p2;
			params.international=p3;

			$( '#paxArriveLocalA' ).val('');
			$( '#paxArriveNonLocalA' ).val('');
			$( '#paxArriveInternationalA' ).val('');
			$( '#paxArriveLocalB' ).val('');
			$( '#paxArriveNonLocalB' ).val('');
			$( '#paxArriveInternationalB' ).val('');

			$.ajax({
				type: "post",
				url: "/ajax/newCabDrop.cfm",
				dataType: "html",
				data: params,
				async: "true",
				cache: "false"})
				.done( function(T)
				{
					$( "#mainbutton" ).hide();
					$( "#altbutton" ).show();
				//	$( "#thecell" ).attr('align', 'right');
					$( "#altbutton" ).animate({
					   // opacity: 0,
					    left: "+=50"
					}, 500, function() {
						$( "#altbutton" ).hide();
						$( "#mainbutton" ).show();
						$( "#thecell" ).attr('align', 'center');
					  });
				})
				.fail( function(xhr, status, errorThrown)
				{
					okToUpdate = true;
					alert(errorThrown);
				})
				.always( function(){
			});


	}

    var viewFullScreen = document.getElementById('view-fullscreen');
    if (viewFullScreen) {
        viewFullScreen.addEventListener('click', function () {
            var docElm = document.documentElement;
            if (docElm.requestFullscreen) {
                docElm.requestFullscreen();
            }
            else if (docElm.msRequestFullscreen) {
                docElm = document.body; //overwrite the element (for IE)
                docElm.msRequestFullscreen();
            }
            else if (docElm.mozRequestFullScreen) {
                docElm.mozRequestFullScreen();
            }
            else if (docElm.webkitRequestFullScreen) {
                docElm.webkitRequestFullScreen();
            }
        }, false);
    }

	document.getElementById('backmenu').addEventListener('click', function(){
	    window.location.href = "/main.cfm";
	});

	var container = document.getElementById('wrapper');
	var inner = document.getElementById('get');
	var inHeight = inner.offsetHeight;
	container.style.height=(window.innerHeight);
	container.style.width=window.innerWidth;
	var conHeight=container.offsetHeight;
	inner.style.marginTop = (conHeight-inHeight)/2+'px';

</script>

**Comment by odysseus55** [comment]
Sorry about that and the delay to get the gist to you.
Briefly, with <cfoutput> on line 164, highlighting is off, move that line
to outside the <script> and works fine.

https://gist.github.com/odysseus55/828a4af9b25e33b6b6fe851e4288ad5d

**Comment by David-Polehonski** [comment]
Great, I've created a test case and will work on it hopefully sometime this week.

**Comment by David-Polehonski** [comment]
The complexities come in when trying to limit syntax highlighting of hash delimited expressions to code blocks within cfoutput tags. It seems hard/impossible to inject the cfoutput into segments of the file where the grammar has already delegated to control to a different grammar. 

language-cfml delegates to language-html. which delegates to language-js or language-css. Trying to get CFML parsing injected at each of the levels without bricking the whole thing is quite difficult. I didn't notice it when I put it together because I've considered it best practice to wrap entire blocks/pages with cfoutput tags to a) keep things tidy, and b) reduce overhead of dropping in and out of cfoutput context.. 

This is going to require a bit more thought...

**Comment by jorabyte** [comment]
@David-Polehonski this issue may have resolved itself.  I was in the process of attaching screenshots of similar syntax highlighting gone wrong, when Atom prompted me to update [atom-ide-ui](https://atom.io/packages/atom-ide-ui) to version 0.10.1, and all of the highlighting issues were fixed for me!  I hope that fixes @odysseus55 issues as well.

**edit:** I'm sorry, the highlighting issues are back.  I don't know why, but when I saw the errors popped up again, I've tried with an without the above-referenced package and still have issues such as:

With language-cfml, script tags aren't being highlighted, **sometimes:**
![image](https://user-images.githubusercontent.com/2533066/38886983-2898b908-423d-11e8-97ee-22aa3b62a8aa.png)

After disabling language-cfml:
![image](https://user-images.githubusercontent.com/2533066/38887085-77582fce-423d-11e8-985c-583da5cc0fdf.png)

With package, code embedded within comments is not shown in comment syntax:
![image](https://user-images.githubusercontent.com/2533066/38887200-c6b286fa-423d-11e8-982b-4bffa0f773d5.png)

Without package:
![image](https://user-images.githubusercontent.com/2533066/38887333-2bbde616-423e-11e8-97b9-f116d2d84bcf.png)

Without language-cfml however, CFC files appear as plain text:
![image](https://user-images.githubusercontent.com/2533066/38887468-8d249846-423e-11e8-86e8-1311bcd4a652.png)

With package:
![image](https://user-images.githubusercontent.com/2533066/38887532-b6f4224a-423e-11e8-90b3-f816f8b55428.png)

**Comment by odysseus55** [comment]
![screenshot from 2018-04-17 11-55-35](https://user-images.githubusercontent.com/952815/38890448-4b54d032-4236-11e8-8811-f330524192cc.png)
I did update to 0.10.1 atom-ide-ui - no change that I can see. 

I took the assignment of CF vars to JS vars outside of the main block. This limits the wonky highlighting to a smaller area. FYI.

**Comment by jorabyte** [comment]
@David-Polehonski : were the highlighting issues only introduced in the roll-out of 0.23.0?  Would you lose much by rolling back to the previous version?  I don't think I had any syntax highlighting issues before the upgrade.

**Comment by David-Polehonski** [comment]
It would need to be a partial rollback as some of the updates address other issues; I'll look into whether a partial reversion would fix the issues.

**Comment by odysseus55** [comment]
I upgraded the atom-ide-ui to version 0.10.1 and issue remains the same.

**Comment by jorabyte** [comment]
Here's why I thought the issue was resolved: **if I have two instances of Atom running at the same time, the 2nd instance has the correct syntax highlighting.**  Let me show you: same file open twice, in two different windows:

![image](https://user-images.githubusercontent.com/2533066/38957051-18cdacee-431f-11e8-853c-ffbfd45ed9be.png)

**Comment by David-Polehonski** [comment]
There seem to be two issues at play here; the first is an intermittent fault that seems to go away after reloading the editor. The other surrounding the cfoutput tag; which is repeatable. 

The latter issue is proving impossible to resolve. The text.html.basic grammar embeds both the css and js grammars within the script and style blocks. So while the cfoutput scope can be applied, it is impossible, it seems, to return to the js or css grammar, I can allow cfml only, or return to html/cfml. It seems that restarting the previous grammar context is not a possibility with the atom grammar. 

What I'm going to propose is removing the cfoutput scope and replacing it with just a simple tag implementation. This would resolve the issue, but then i'd either have to drop hash delimited expressions or enable them for the whole file. The purpose of the cfoutput scope was to only enable them within an output context.

**Comment by MrJimOT** [comment]
The second pound sign in the href line changes the syntax color in Atom. Not a huge problem. I just leave the cffunction containing the line last in the 7000 line .cfc. More FYI.

![image](https://user-images.githubusercontent.com/23161206/45255663-68646b80-b34f-11e8-96fe-7162c6e9ae29.png)
